tests/unit_tests/test_protocol.py

```diff
@@ -256,3 +259,117 @@ async def test_read_when_disconnected(protocol):
 
     with pytest.raises(DisconnectedError):
         await protocol.read_message()
+
+
+def test_json_encoder_float_serialization():
+    assert json_encoder.encode(123.0) == '123.0'
+    assert json_encoder.encode(0.99) == '0.99'
+    assert json_encoder.encode(0.999) == '1.0'
+
+
+def test_json_encoder_encodes_server_messages():
+    new_encode = json_encoder.encode
+    old_encode = json.JSONEncoder(separators=(",", ":")).encode
+
+    delta_ladder, delta_tmm = 200.3654233558, 19.879223123
+
+    matchmaker_info_dict = {
+        "command": "matchmaker_info",
+        "queues": [
+            {
+                "queue_name": "ladder1v1",
+                "queue_pop_time": "2002-01-01T23:59:59.595959",
+                "queue_pop_time_delta": delta_ladder,
+                "num_players": 3,
+                "boundary_80s": [(1200, 1600), (1600, 2000), (300, 700)],
+                "boundary_75s": [(1300, 1500), (1700, 1900), (400, 600)],
+                "team_size": 1,
+            },
+            {
+                "queue_name": "tmm2v2",
+                "queue_pop_time": "2002-01-01T23:56:23.234255",
+                "queue_pop_time_delta": delta_tmm,
+                "num_players": 4,
+                "boundary_80s": [
+                    (1200, 1600), (1600, 2000), (300, 700), (300, 700),
+                ],
+                "boundary_75s": [
+                    (1300, 1500), (1700, 1900), (400, 600), (400, 600),
+                ],
+                "team_size": 2,
+            },
+        ]
+    }
+
+    expected_matchmaker_info_dict = copy.deepcopy(matchmaker_info_dict)
+
+    for queue in expected_matchmaker_info_dict["queues"]:
+        queue["queue_pop_time_delta"] = round(
+            queue["queue_pop_time_delta"], config.JSON_FLOAT_MAX_DIGITS,
+        )
+
+    assert new_encode(matchmaker_info_dict) == (
+        old_encode(expected_matchmaker_info_dict)
+    )
+
+    game_info_dict = {
+        "command": "game_info",
+        "visibility": "public",
+        "password_protected": True,
+        "uid": 13,
+        "title": "someone's game",
+        "state": "playing",
+        "game_type": "custom",
+        "featured_mod": "faf",
+        "sim_mods": {},
+        "mapname": "scmp_009",
+        "map_file_path": "maps/scmp_009.zip",
+        "host": "Foo",
+        "num_players": 2,
+        "launched_at": 1111111111.11,
+        "rating_type": "faf",
+        "rating_min": None,
+        "rating_max": None,
+        "enforce_rating_range": False,
+        "team_ids": [
+            {
+                "team_id": 1,
+                "player_ids": [1],
+            },
+            {
+                "team_id": 2,
+                "player_ids": [2],
+            },
+        ],
+        "teams": {
+            1: ["Foo"],
+            2: ["Bar"],
+        },
+    }
+    assert new_encode(game_info_dict) == old_encode(game_info_dict)
+
+
+def test_json_encoder_encodes_dicts():
```

Comment:
This could be a good test to write using `hypothesis`. You would just need to make a composite strategy that can generate any of the type's you're testing (bools, None, strings, ints, floats). Hypothesis will take care of figuring out the actual data.